File: Ashutosh-Singh/BinarySearch/Magnetic-Force-Between-Two-Balls.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    bool isPlacement(vector<int>& position, int m, int mid) {
        int count = 1;
        int last = position[0];
        for (int i = 0; i < position.size(); i++) {
            if (position[i] - last >= mid) {
                last = position[i];
                count++;
            }
        }
        return count >= m;
    }

    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(), position.end());
        int n = position.size();
        int st = 0;
        int end = position[n - 1];
        int mid;
        int optimal = 0;
        while (st <= end) {
            mid = st + (end - st) / 2;
            if (isPlacement(position, m, mid)) {
                optimal = mid;
                st = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return optimal;
    }
};

int main() {
    Solution solution;
    int n, m;
    cout << "Enter the number of positions: ";
    cin >> n;

    vector<int> position(n);
    cout << "Enter the positions: ";
    for (int i = 0; i < n; i++) {
        cin >> position[i];
    }

    cout << "Enter the number of balls to place: ";
    cin >> m;

    int result = solution.maxDistance(position, m);
    cout << "The maximum distance between any two balls is: " << result << endl;

    return 0;
}

File: Anya_Seth/BinarySearch/Magnetic Force Between Two Balls.java
class Solution {
    public int maxDistance(int[] position, int m){
        Arrays.sort(position);
        int min = 1;
        int max = position[position.length - 1] - position[0];
        while(min <= max){
            int mid = (min + max)/2;
            if(isValid(position, mid, m)){
                min = mid + 1;
            }
            else{
                max = mid - 1;
            }
        }
        return max;
    }
    public static boolean isValid(int[] position, int mid, int m){
        int count = 1;
        int first_pos = position[0];
        for(int i = 1 ; i < position.length ; i++){
           if(position[i] - first_pos >= mid){
             first_pos = position[i];
             count++;
            }
        }
        return count >= m;
    }
}
